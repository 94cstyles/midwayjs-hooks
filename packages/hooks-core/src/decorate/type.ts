import type { Merge } from 'type-fest'
import { AsyncFunction } from '../'

export type ArrayToObject<T, R = {}> = T extends [infer First, ...infer Rest]
  ? First extends PromiseLike<infer PromiseValue>
    ? PromiseValue
    : First extends object
    ? Merge<First, ArrayToObject<Rest, R>>
    : ArrayToObject<Rest, R>
  : R

export type DecorateHandler<
  Input extends object | void,
  Handler extends AsyncFunction
> = (
  ...args: Input extends void
    ? Parameters<Handler>
    : [input: Input, ...args: Parameters<Handler>]
) => ReturnType<Handler>

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export type DefineHelper = {
  setProperty: (key: any, value: any) => void
  getProperty: (key: any) => any
}

export type ExecuteHelper = {
  result?: any
  next?: () => Promise<void>
  getInputArguments?: () => any[]
}

export type Operator<Input> = {
  name: string
  type?: Input
  requireInput?: boolean
  defineMeta?: (helper: DefineHelper) => void
  execute?: (helper: ExecuteHelper) => Promise<void>
}

export type ExtractInputType<T> = {
  [key in keyof T]: T[key] extends Operator<any> ? T[key]['type'] : void
}
