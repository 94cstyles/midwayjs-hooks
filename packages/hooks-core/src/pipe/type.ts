import type { Merge } from 'type-fest'

export type ArrayToObject<T, R = {}> = T extends [infer First, ...infer Rest]
  ? First extends PromiseLike<infer PromiseValue>
    ? PromiseValue
    : First extends object
    ? Merge<First, ArrayToObject<Rest, R>>
    : ArrayToObject<Rest, R>
  : R

export type PipeHandler<
  Input extends object | void,
  Handler extends (...args: any) => Promise<any>
> = (
  ...args: Input extends void
    ? Parameters<Handler>
    : [input: Input, ...args: Parameters<Handler>]
) => ReturnType<Handler>

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export type DefineHelper = {
  setProperty: (key: any, value: any) => void
  getProperty: (key: any) => any
}

type ExecuteHelper = {
  next?: () => Promise<void>
}

export type Operator<Input> = {
  name: string
  type?: Input
  requireInput?: boolean
  defineMeta?: (helper: DefineHelper) => void
  execute?: (helper: ExecuteHelper) => Promise<void>
}

export type ExtractInputType<T> = {
  [key in keyof T]: T[key] extends Operator<any> ? T[key]['type'] : void
}
