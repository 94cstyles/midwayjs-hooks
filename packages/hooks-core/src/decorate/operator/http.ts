import { DefineHelper, Operator, OperatorType } from '../type'

export const HttpTrigger = 'HTTP'

export enum HttpMetadata {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAM = 'Http_Param',
  HEADER = 'Http_Header',
  RESPONSE = 'Http_Response',
}

export enum ResponseMetadata {
  CODE = 'Http_Response_Code',
  HEADER = 'Http_Response_Header',
  CONTENT_TYPE = 'Http_Response_ContentType',
  REDIRECT = 'Http_Response_Redirect',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      metadata({ setMetadata }) {
        setMetadata(OperatorType.Trigger, {
          type: HttpTrigger,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Delete = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpMetadata.QUERY,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpMetadata.PARAM,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpMetadata.HEADER,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.HEADER, true)
    },
  }
}

export type ResponseMetaData = {
  type: ResponseMetadata
  code?: number
  header?: {
    key: string
    value: string
  }
  url?: string
  contentType?: string
}

export function HttpCode(code: number): Operator<void> {
  return {
    name: 'HttpCode',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.CODE, { code })
    },
  }
}

export function SetHeader(key: string, value: string): Operator<void> {
  return {
    name: 'SetHeader',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.HEADER, {
        header: {
          key,
          value,
        },
      })
    },
  }
}

export function Redirect(url: string, code?: number): Operator<void> {
  return {
    name: 'Redirect',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.REDIRECT, {
        url,
        code,
      })
    },
  }
}

export function ContentType(contentType: string): Operator<void> {
  return {
    name: 'ContentType',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetadata.CONTENT_TYPE, {
        contentType,
      })
    },
  }
}

function setResponseMetaData(
  helper: DefineHelper,
  type: ResponseMetadata,
  value: Partial<ResponseMetaData>
) {
  const responseMetaData =
    helper.getMetadata<ResponseMetaData[]>(HttpMetadata.RESPONSE) || []

  helper.setMetadata(HttpMetadata.RESPONSE, [
    ...responseMetaData,
    {
      type,
      ...value,
    },
  ])
}
