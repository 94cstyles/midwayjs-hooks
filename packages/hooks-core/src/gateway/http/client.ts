import { EXPORT_DEFAULT_FUNCTION_ALIAS } from '../../const'
import { createApiClientMatcher } from '../../index'
import { HTTPRouter } from './router'

function getUrl(baseUrl: string, name: string) {
  if (name === EXPORT_DEFAULT_FUNCTION_ALIAS) {
    return baseUrl
  }

  if (baseUrl.endsWith('/')) {
    return `${baseUrl}${name}`
  }

  return `${baseUrl}/${name}`
}

export function createHTTPClientMatcher(router: HTTPRouter) {
  return createApiClientMatcher().match((meta) => {
    const { file, route, functionName } = meta

    if (!route.basePath) {
      return
    }

    const baseUrl = router.getBaseUrl(file)
    const url = getUrl(baseUrl, functionName)
    const client =
      router.projectConfig?.request?.client || '@midwayjs/hooks/request'

    return {
      client,
      metadata: {
        url,
        meta: {},
      },
    }
  })
}
