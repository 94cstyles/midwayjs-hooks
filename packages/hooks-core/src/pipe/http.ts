import { Operator, OperatorProperty } from './type'

const Protocol = 'HTTP'

enum HttpProperty {
  METHOD = 'METHOD',
  QUERY = 'QUERY',
  PARAM = 'PARAM',
  HEADER = 'HEADER',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      defineMeta({ setProperty: defineProperty }) {
        defineProperty(OperatorProperty.Trigger, {
          type: Protocol,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Delete = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpProperty.QUERY,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpProperty.PARAM,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpProperty.HEADER,
    defineMeta({ setProperty: defineProperty }) {
      defineProperty(HttpProperty.HEADER, true)
    },
  }
}
