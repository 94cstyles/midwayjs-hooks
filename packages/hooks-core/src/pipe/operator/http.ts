import { Operator, OperatorType } from '../type'

export const HttpTrigger = 'HTTP'

export enum HttpProperty {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAM = 'Http_Param',
  HEADER = 'Http_Header',
}

function createHTTPMethodOperator(method: string) {
  return () => {
    return {
      name: method,
      defineMeta({ setProperty }) {
        setProperty(OperatorType.Trigger, {
          type: HttpTrigger,
          method,
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHTTPMethodOperator('ALL')
export const Get = createHTTPMethodOperator('GET')
export const Post = createHTTPMethodOperator('POST')
export const Put = createHTTPMethodOperator('PUT')
export const Del = createHTTPMethodOperator('DELETE')
export const Patch = createHTTPMethodOperator('PATCH')
export const Head = createHTTPMethodOperator('HEAD')
export const Options = createHTTPMethodOperator('OPTIONS')

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpProperty.QUERY,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.QUERY, true)
    },
  }
}

export function Param<T extends Record<string, string>>(): Operator<{
  param: T
}> {
  return {
    name: HttpProperty.PARAM,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.PARAM, true)
    },
  }
}

export function Header<T extends Record<string, string>>(): Operator<{
  header: T
}> {
  return {
    name: HttpProperty.HEADER,
    requireInput: true,
    defineMeta({ setProperty }) {
      setProperty(HttpProperty.HEADER, true)
    },
  }
}
