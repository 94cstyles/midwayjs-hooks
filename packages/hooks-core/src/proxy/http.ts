import { proxy } from './proxy'
import { request } from '../request'
import { getHTTPMethod } from '../router'

export function createRequestProxy(baseUrl: string) {
  return proxy(async (func, thisArg, args) => {
    const name = func._param.meta.name
    return request({
      url: baseUrl + name,
      method: getHTTPMethod(args.length),
      data: { args },
      meta: {},
    })
  })
}
