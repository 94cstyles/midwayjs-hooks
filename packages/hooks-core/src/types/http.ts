/**
 * Now the function accepts all http methods
 * @deprecated will be removed in next major version.
 */
export type ApiHttpMethod = 'GET' | 'POST'

export type ApiBaseParam = {
  /**
   * @description generated by the file router, base on file path
   */
  functionId?: string
}

export interface ApiParam extends ApiBaseParam {
  url: string
  /**
   * Now the function accepts all http methods
   * @deprecated will be removed in next major version.
   */
  method: ApiHttpMethod
  data: {
    args: any[]
    [key: string]: any
  }
  meta: {
    functionName?: string
    superjson?: boolean
    [key: string]: any
  }
}
