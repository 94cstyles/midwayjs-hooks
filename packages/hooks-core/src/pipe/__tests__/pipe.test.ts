import { Middleware, extractMetadata } from '@midwayjs/hooks-core'

import { Pipe } from '../'
import { Get, Header, Param, Post, Query } from '../operator/http'
import { Operator } from '../type'

it('pipe should work', async () => {
  const hello = Pipe(async () => {
    return 'Hello Pipe!'
  })
  expect(await hello()).toEqual('Hello Pipe!')

  const withArgs = Pipe(async (name: string) => {
    return name
  })
  expect(await withArgs('hooks')).toEqual('hooks')

  const withOperator = Pipe(Post(), async (name: string) => {
    return name
  })
  expect(await withOperator('hooks')).toEqual('hooks')
})

it('compose with meta operator', async () => {
  const withQuery = Pipe(Get(), Query<{ name: string }>(), async () => {
    return 'Hello World'
  })
  expect(await withQuery({ query: { name: 'hooks' } })).toEqual('Hello World')
})

it('defineMeta', async () => {
  const logger = () => {}
  const fn = Pipe(
    Get(),
    Query(),
    Param(),
    Header(),
    Middleware(logger),
    async () => {}
  )

  expect(extractMetadata(fn)).toMatchSnapshot()
})

it('execute', async () => {
  const stack = []
  const CustomExecutor = (): Operator<void> => {
    return {
      name: 'CustomExecutor',
      async execute({ next }) {
        stack.push(1)
        await next()
        stack.push(4)
      },
    }
  }

  const fn = Pipe(CustomExecutor(), async (...numbers: number[]) => {
    stack.push(...numbers)
    return numbers
  })

  const result = await fn(2, 3)
  expect(stack).toEqual([1, 2, 3, 4])
  expect(result).toEqual([2, 3])
})
