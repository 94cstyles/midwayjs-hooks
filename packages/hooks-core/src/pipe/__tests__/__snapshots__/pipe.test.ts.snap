// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineMeta 1`] = `
Object {
  "Http_Header": true,
  "Http_Param": true,
  "Http_Query": true,
  "Middleware": Array [
    [Function],
  ],
  "Trigger": Object {
    "method": "GET",
    "type": "HTTP",
  },
  "isPipe": true,
}
`;
